common/djangoapps/student/tests/tests.py

```diff
@@ -1145,3 +1169,133 @@ def _assert_responses(self, response, count):
+            self.assertFalse(mock_method.called)
+            self.assert_parsed_data(parse_data2)
+
+            # Reset the cache.
+            cache.clear()
+            parse_data3 = _get_course_programs(self.user, course_list)
+            mock_method.assert_called_once_with(self.user, course_list)
+            self.assertTrue(mock_method.called)
+            self.assert_parsed_data(parse_data3)
+
+    def test_set_programs_cache_with_empty_dict(self):
+        """Verify that cache will be set even in case of empty dict."""
+        self.create_config(enabled=True, enable_student_dashboard=True, enable_cache=True, cache_ttl=100)
+        course_id = u'edx/demox/Run_1'
+        del self.raw_data[u'edx/demox/Run_1'][u'name']
+
+        with patch('student.views.get_course_programs_for_dashboard') as mock_method:
+            mock_method.return_value = self.raw_data
+
+            # First call will warm them cache
+            parse_data1 = _get_course_programs(self.user, [course_id])
+            mock_method.assert_called_once_with(self.user, [course_id])
+            self.assertTrue(mock_method.called)
+            self.assertEqual({}, parse_data1)
+
+            mock_method.reset_mock()
+
+            # For second attempt cache will return the data.
+            # Method will not be called
+            parse_data2 = _get_course_programs(self.user, [course_id])
+            self.assertFalse(mock_method.called)
+            self.assertEqual({}, parse_data2)
+
+            # Reset the cache.
+            cache.clear()
+            parse_data3 = _get_course_programs(self.user, [course_id])
+            mock_method.assert_called_once_with(self.user, [course_id])
+            self.assertTrue(mock_method.called)
+            self.assertEqual({}, parse_data3)
+
+    def assert_parsed_data(self, parsed_data):
+        """Dry method to compare responses."""
+        self.assertEqual(self.parsed_data, parsed_data)
+
+    def test_program_courses_on_dashboard_with_cache(self):
+        """Verify that if programs config has cache enabled than on dashboard
+         page _get_course_programs method will cache the response.
+        """
+        CourseEnrollment.enroll(self.user, self.course_1.id, mode='verified')
+
+        self.client.login(username="jack", password="test")
+        self.create_config(enabled=True, enable_student_dashboard=True, enable_cache=True, cache_ttl=100)
+
+        with patch('student.views.get_course_programs_for_dashboard') as mock_method:
+            mock_method.return_value = self._create_program_data(
+                [(self.course_1.id, 'active'), (self.course_2.id, 'unpublished')]
+            )
+
+            response = self.client.get(reverse('dashboard'))
+            mock_method.assert_called_once_with(self.user, [
+                self.course_1.id
+            ])
+            self.assertTrue(mock_method.called)
+            # count total courses appearing on student dashboard
```

Comment:
nit: be consistent with using either small letter or camel case for add comments inside method